tests/sync/app.cpp

```diff
@@ -1627,3 +1627,134 @@ TEST_CASE("app: remove user with credentials", "[sync][app]") {
     }
     
 }
+
+TEST_CASE("app: link_user", "[sync][app]") {
+
+    SECTION("link_user") {
+        std::unique_ptr<GenericNetworkTransport> (*transport_generator)() = []{
+            struct transport : GenericNetworkTransport {
+                void send_request_to_server(const Request request,
+                                            std::function<void (const Response)> completion_block)
+                {
+                    if (request.url.find("/login") != std::string::npos) {
+                        completion_block({
+                            200, 0, {}, user_json(good_access_token).dump()
+                        });
+                    } else if (request.url.find("/profile") != std::string::npos) {
+                        completion_block({
+                            200, 0, {}, user_profile_json().dump()
+                        });
+                    } else if (request.url.find("/session") != std::string::npos) {
+                        CHECK(request.method == HttpMethod::del);
+                        completion_block({ 200, 0, {}, "" });
+                    }
+                }
+            };
+            return std::unique_ptr<GenericNetworkTransport>(new transport);
+        };
+        
+        auto config = App::Config{"translate-utwuv", transport_generator};
+        std::string base_path = tmp_dir() + "/" + config.app_id;
+        reset_test_directory(base_path);
+        auto tsm = TestSyncManager(base_path);
+        
+        auto app = App(config);
+
+        bool processed = false;
+
+        std::shared_ptr<SyncUser> sync_user;
+        
+        auto email = util::format("realm_tests_do_autoverify%1@%2.com", random_string(10), random_string(10));
+        auto password = random_string(10);
+        
+        auto custom_credentials = realm::app::AppCredentials::facebook("a_token");
+        auto email_pass_credentials = realm::app::AppCredentials::username_password(email, password);
+
+        app.log_in_with_credentials(email_pass_credentials,
+                                    [&](std::shared_ptr<realm::SyncUser> user, Optional<app::AppError> error) {
+            REQUIRE(user);
+            CHECK(!error);
+            sync_user = user;
+        });
+        
+        CHECK(sync_user->provider_type() == IdentityProviderUsernamePassword);
+
+        app.link_user(sync_user,
+                      custom_credentials,
+                      [&](std::shared_ptr<SyncUser> user, Optional<app::AppError> error) {
+            CHECK(!error);
+            REQUIRE(user);
+            CHECK(user->identity() == sync_user->identity());
```

Comment:
The doc for this method says that a new user is returned, but this test seems to contradict this as they have a similar identity and provider_type?

Reply:
original impl was wrong, fixed to take a new user id